### PR TITLE
Prometheus: Simplify HTTP client(s)

### PR DESCRIPTION
**Pull request by toddtreece**
**What this PR does / why we need it**:

This uses changes from #50466 to simplify the Prometheus HTTP client(s).

Notes:
* the current state hasn't been tested, but `client`, `resource` and `querydata` should mostly be migrated
* `buffered` and `buffered/promclient` still need to be migrated

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/query/query.go [hunk 1]
```diff
@@ -136,6 +138,13 @@ func (s *Service) handleQueryData(ctx context.Context, user *models.SignedInUser
 		Queries: []backend.DataQuery{},
 	}
 
+	middlewares := []httpclient.Middleware{}
+	if parsedReq.httpRequest != nil {
```
Comment (aocenas) [inline pkg/services/query/query.go:142]: What is the case when `parsedReq.httpRequest` is nil? seems like things should be HTTP requests here.

File: pkg/infra/httpclient/httpclientprovider/forwarded_cookies_middleware.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
+	"github.com/grafana/grafana/pkg/util/proxyutil"
+)
+
+const ForwardedCookiesMiddlewareName = "forwarded-cookies"
+
+// ForwardedCookiesMiddleware middleware that sets Cookie header on the
+// outgoing request, if forwarded cookies configured/provided.
+func ForwardedCookiesMiddleware(forwardedCookies []*http.Cookie, allowedCookies []string) httpclient.Middleware {
```
Comment (aocenas) [inline pkg/infra/httpclient/httpclientprovider/forwarded_cookies_middleware.go:14]: Feels like needing to do the filtering here mixes concerns. Should be easy to just pass on cookies to forward and have a util function to do the filtering before in the caller code.

File: pkg/services/query/query.go [hunk 3]
```diff
@@ -144,6 +153,7 @@ func (s *Service) handleQueryData(ctx context.Context, user *models.SignedInUser
 			if ok && idToken != "" {
 				req.Headers["X-ID-Token"] = idToken
```
Comment (aocenas) [inline pkg/services/query/query.go:154]: Seems like this code is now duplicated here and in the `ForwardedOAuthIdentityMiddleware`

File: pkg/tsdb/prometheus/client/client.go [hunk 4]
```diff
@@ -13,100 +13,92 @@ import (
...
 
-func NewClient(d doer, method, baseUrl string) *Client {
-	return &Client{doer: d, method: method, baseUrl: baseUrl}
+func NewClient(rt http.RoundTripper, method, baseUrl string) *Client {
+	return &Client{RoundTripper: rt, method: method, baseUrl: baseUrl}
 }
 
 func (c *Client) QueryRange(ctx context.Context, q *models.Query) (*http.Response, error) {
-	u, err := url.ParseRequestURI(c.baseUrl)
+	u, err := url.Parse("/api/v1/query_range")
```
Comment (aocenas) [inline pkg/tsdb/prometheus/client/client.go:27]: Why is this change here does not seem like it's entirely connected to the other changes.

File: pkg/tsdb/prometheus/client/provider.go [hunk 5]
```diff
@@ -1,93 +0,0 @@
...
-		return nil, err
-	}
-
-	return NewClient(httpClient, p.httpMethod, p.settings.URL), nil
-}
-
-func (p *Provider) middlewares() []sdkhttpclient.Middleware {
-	middlewares := []sdkhttpclient.Middleware{
-		middleware.CustomQueryParameters(p.log),
-		sdkhttpclient.CustomHeadersMiddleware(),
```
Comment (aocenas) [inline pkg/tsdb/prometheus/client/provider.go:81]: Don't we need these middlewares anymore?

Comment (toddtreece) [reply pkg/tsdb/prometheus/client/provider.go:81]: any middleware that don't already get added in `pkg/infra/httpclient` could be added as contextual middleware used by both prom buffered and streaming code paths. this file was deleted because prometheus shouldn't need its own httpclients after things have been migrated to contextual middleware

File: pkg/services/query/query.go (hunk 1 above)
Comment (toddtreece) [reply pkg/services/query/query.go]: this branch was pointed at https://github.com/grafana/grafana/pull/50466, so these changes aren't related to this PR

File: pkg/services/query/query.go (hunk 3 above)
Comment (aocenas) [reply pkg/services/query/query.go]: But if add the middleware in the next few lines doesn't it mean that all data sources will use that middleware? Like what is the point of data source even be aware of the middleware?

File: pkg/tsdb/prometheus/client/provider.go (hunk 5 above)
Comment (aocenas) [reply pkg/tsdb/prometheus/client/provider.go:81]: So does this mean these middlewares are already added in pkg/infra/httpclient? I see the `CustomHeadersMiddleware` but not `CustomQueryParameters`.

File: pkg/infra/httpclient/httpclientprovider/forwarded_cookies_middleware.go (hunk 2 above)
Comment (aocenas) [reply pkg/infra/httpclient/httpclientprovider/forwarded_cookies_middleware.go]: hmm and can we change it anyway? This signature seems to me like `addNumberButFilterOutOneNumber(numbers []int, withoutNumber int)`. Like it seems like it should be much easier to just compose it like `addNumbers(numbers.filter(numberToFilter))` or something like that.

Comment (toddtreece) [reply pkg/infra/httpclient/httpclientprovider/forwarded_cookies_middleware.go]: @aocenas oh yeah, for sure. it can be changed if you have ideas for improvements. i'd recommend doing that in a separate PR so it's easier to get feedback from `@grafana/plugins-platform-backend` without the prometheus changes mixed in
